tokenizers/src/tokenizer/mod.rs

```diff
@@ -845,22 +847,35 @@ where
 
     /// Decode the given ids, back to a String
     pub fn decode(&self, ids: &[u32], skip_special_tokens: bool) -> Result<String> {
-        let tokens = ids
-            .iter()
-            .filter_map(|id| {
-                self.added_vocabulary
-                    .id_to_token(*id, &self.model)
-                    .filter(|token| {
-                        !skip_special_tokens || !self.added_vocabulary.is_special_token(token)
-                    })
-            })
-            .collect::<Vec<_>>();
-
-        if let Some(decoder) = &self.decoder {
-            decoder.decode(tokens)
-        } else {
-            Ok(tokens.join(" "))
+        let mut result = String::with_capacity(ids.len());
+        let mut chunks = Vec::with_capacity(ids.len());
+        for id in ids {
+            if let Some(added_token) = self.added_vocabulary.simple_id_to_token(*id) {
+                if skip_special_tokens && self.added_vocabulary.is_special_token(&added_token) {
```

Comment:
```suggestion
                if skip_special_tokens && added_token.special) {
```
should be more foolproof?

Reply:
actually no worries